File: stations/agent-metrics/src/agent_metrics/routes/sysmon.py
"""Sysmon + Quota + Guardian/Sweep API routes."""

from __future__ import annotations

from fastapi import APIRouter, Query

from agent_metrics.db import get_pool
from agent_metrics.quota_collector import (
    get_quota,
    get_quota_health,
    get_raw_cache,
    reset_cc_backoff,
)
from agent_metrics.sysmon_loop import get_history, get_latest

router = APIRouter()


# ---------------------------------------------------------------------------
# System Metrics
# ---------------------------------------------------------------------------


@router.get("/sysmon/current")
async def sysmon_current():
    """Return the latest system metrics snapshot."""
    data = get_latest()
    if not data:
        return {"error": "no data yet"}
    return data


@router.get("/sysmon/history")
async def sysmon_history(minutes: int = Query(default=60, ge=1, le=720)):
    """Return sysmon history from ring buffer."""
    return {"entries": get_history(minutes), "interval_s": 5}


# ---------------------------------------------------------------------------
# LLM Quota
# ---------------------------------------------------------------------------


@router.get("/quota/current")
async def quota_current():
    """Return raw + parsed quota data for all LLM providers."""
    quota = await get_quota()
    raw = get_raw_cache()
    health = get_quota_health()
    return {
        "raw": raw,
        "health": health,
        "parsed": {
            "cc": quota.get("cc_parsed", {}),
            "cx": quota.get("cx_parsed", {}),
            "gm": quota.get("gm_parsed", {}),
        },
        "formatted": {
            "cc_5h": quota.get("llm_cc_5h", "?"),
            "cc_7d": quota.get("llm_cc_7d", "?"),
            "cc_ex": quota.get("llm_cc_ex", "?"),
            "cx_5h": quota.get("llm_cx_5h", "?"),
            "cx_7d": quota.get("llm_cx_7d", "?"),
            "gm_pro": quota.get("llm_gm_pro", "?"),
            "gm_flash": quota.get("llm_gm_flash", "?"),
        },
    }


@router.get("/quota/formatted")
async def quota_formatted():
    """Return tmux-compatible formatted quota strings."""
    quota = await get_quota()
    return {
        "cc-5h": quota.get("llm_cc_5h", "?"),
        "cc-7d": quota.get("llm_cc_7d", "?"),
        "cc-ex": quota.get("llm_cc_ex", "?"),
        "cx-5h": quota.get("llm_cx_5h", "?"),
        "cx-7d": quota.get("llm_cx_7d", "?"),
        "gm-pro": quota.get("llm_gm_pro", "?"),
        "gm-flash": quota.get("llm_gm_flash", "?"),
        "display": quota.get("llm_display", "?"),
    }


@router.post("/quota/reset-cc-backoff")
async def quota_reset_cc_backoff():
    """Reset CC backoff and force a fresh fetch on next tick."""
    reset_cc_backoff()
    quota = await get_quota(force=True)
    health = get_quota_health()
    return {
        "action": "backoff_reset",
        "health": health.get("cc", {}),
        "cc_5h": quota.get("llm_cc_5h", "?"),
        "cc_7d": quota.get("llm_cc_7d", "?"),
    }


# ---------------------------------------------------------------------------
# Guardian / Sweep Logs
# ---------------------------------------------------------------------------


@router.get("/guardian/log")
async def guardian_log(
    hours: int = Query(default=24, ge=1, le=168),
    level: str | None = Query(default=None),
):
    """Return guardian action log."""
    pool = await get_pool()
    query = (
        "SELECT id, ts, level, priority, pid, process_name, mem_mb, cpu_pct, "
        "action, result, detail FROM guardian_actions "
        "WHERE ts > now() - make_interval(hours => $1) AND level != 'SWEEP'"
    )
    params: list = [hours]
    if level:
        query += " AND level = $2"
        params.append(level)
    query += " ORDER BY ts DESC LIMIT 200"

    rows = await pool.fetch(query, *params)
    actions = [dict(r) for r in rows]
    for a in actions:
        if a.get("ts"):
            a["ts"] = a["ts"].isoformat()
    return {"actions": actions, "total": len(actions)}


@router.get("/sweep/log")
async def sweep_log(hours: int = Query(default=24, ge=1, le=168)):
    """Return sweep action log."""
    pool = await get_pool()
    rows = await pool.fetch(
        "SELECT id, ts, level, priority, pid, process_name, mem_mb, cpu_pct, "
        "action, result, detail FROM guardian_actions "
        "WHERE ts > now() - make_interval(hours => $1) AND level = 'SWEEP' "
        "ORDER BY ts DESC LIMIT 200",
        hours,
    )
    actions = [dict(r) for r in rows]
    for a in actions:
        if a.get("ts"):
            a["ts"] = a["ts"].isoformat()
    return {"actions": actions, "total": len(actions)}

File: stations/agent-metrics-rs/src/web/sysmon.rs
//! /sysmon/* and /quota/* routes.

use super::AppState;
use axum::{
    extract::{Query, State},
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};

#[derive(Deserialize)]
pub struct HistoryQuery {
    #[serde(default = "default_minutes")]
    pub minutes: i64,
}
fn default_minutes() -> i64 {
    60
}

pub async fn current(State(state): State<AppState>) -> Json<Value> {
    let latest = state.loop_state.latest.read().await.clone();
    match latest {
        Some(snap) => Json(serde_json::to_value(snap).unwrap_or(Value::Null)),
        None => Json(json!({"error": "no data yet"})),
    }
}

pub async fn history(
    State(state): State<AppState>,
    Query(q): Query<HistoryQuery>,
) -> Json<Value> {
    let interval_s = state.settings.sysmon_collect_interval as i64;
    let max_entries = if q.minutes <= 0 || interval_s <= 0 {
        0
    } else {
        ((q.minutes * 60) / interval_s) as usize
    };
    let buf = state.loop_state.history.read().await;
    let take = buf.len().min(max_entries);
    let entries: Vec<_> = buf.iter().rev().take(take).collect::<Vec<_>>().into_iter().rev().collect();
    let entries_json: Vec<Value> = entries.iter().map(|s| serde_json::to_value(s).unwrap_or(Value::Null)).collect();
    Json(json!({"entries": entries_json, "interval_s": interval_s}))
}

pub async fn quota_current(State(state): State<AppState>) -> Json<Value> {
    let formatted = crate::collectors::quota::get_quota(&state.settings).await;
    let raw = crate::collectors::quota::get_raw_cache(&state.settings).await;
    Json(json!({
        "raw": raw,
        "health": Value::Null,
        "parsed": {"cc": {}, "cx": {}, "gm": {}},
        "formatted": {
            "cc_5h": formatted.llm_cc_5h,
            "cc_7d": formatted.llm_cc_7d,
            "cc_ex": formatted.llm_cc_ex,
            "cx_5h": formatted.llm_cx_5h,
            "cx_7d": formatted.llm_cx_7d,
            "gm_pro": formatted.llm_gm_pro,
            "gm_flash": formatted.llm_gm_flash,
            "cc_5h_resets_at": formatted.llm_cc_5h_resets_at,
            "cc_7d_resets_at": formatted.llm_cc_7d_resets_at,
            "cx_5h_resets_at": formatted.llm_cx_5h_resets_at,
            "cx_7d_resets_at": formatted.llm_cx_7d_resets_at,
            "gm_daily_resets_at": formatted.llm_gm_daily_resets_at,
            "cc_ex_used_usd": formatted.llm_cc_ex_used_usd,
            "cc_ex_limit_usd": formatted.llm_cc_ex_limit_usd,
            "cc_ex_balance_usd": formatted.llm_cc_ex_balance_usd,
            "cc_ex_utilization": formatted.llm_cc_ex_utilization,
            "cc_ex_enabled": formatted.llm_cc_ex_enabled,
        }
    }))
}

pub async fn quota_formatted(State(state): State<AppState>) -> Json<Value> {
    let q = crate::collectors::quota::get_quota(&state.settings).await;
    Json(json!({
        "cc-5h": q.llm_cc_5h,
        "cc-7d": q.llm_cc_7d,
        "cc-ex": q.llm_cc_ex,
        "cx-5h": q.llm_cx_5h,
        "cx-7d": q.llm_cx_7d,
        "gm-pro": q.llm_gm_pro,
        "gm-flash": q.llm_gm_flash,
        "display": q.llm_display,
    }))
}
